arangod/Aql/ClusterQuery.cpp

```diff
@@ -45,43 +46,44 @@ using namespace arangodb::aql;
 ClusterQuery::ClusterQuery(QueryId id,
                            std::shared_ptr<transaction::Context> ctx,
                            QueryOptions options)
-    : Query(id, ctx, aql::QueryString(),
-            /*bindParams*/ nullptr, std::move(options),
+    : Query{id,
+            ctx,
+            {},
+            /*bindParams*/ nullptr,
+            std::move(options),
             /*sharedState*/ ServerState::instance()->isDBServer()
                 ? nullptr
-                : std::make_shared<SharedQueryState>(ctx->vocbase().server())) {
-}
-
-ClusterQuery::~ClusterQuery() {
-  try {
-    _traversers.clear();
-  } catch (...) {
-  }
+                : std::make_shared<SharedQueryState>(ctx->vocbase().server())} {
 }
 
 /// @brief factory method for creating a cluster query. this must be used to
 /// ensure that ClusterQuery objects are always created using shared_ptrs.
-/*static*/ std::shared_ptr<ClusterQuery> ClusterQuery::create(
+std::shared_ptr<ClusterQuery> ClusterQuery::create(
     QueryId id, std::shared_ptr<transaction::Context> ctx,
-    aql::QueryOptions options) {
-  // workaround to enable make_shared on a class with a private/protected
-  // constructor
-  struct MakeSharedQuery : public ClusterQuery {
+    QueryOptions options) {
+  // workaround to enable make_shared on a class with a protected constructor
+  struct MakeSharedQuery final : ClusterQuery {
     MakeSharedQuery(QueryId id, std::shared_ptr<transaction::Context> ctx,
-                    aql::QueryOptions options)
-        : ClusterQuery(id, std::move(ctx), std::move(options)) {}
+                    QueryOptions options)
+        : ClusterQuery{id, std::move(ctx), std::move(options)} {}
+
+    ~MakeSharedQuery() final {
+      _queryProfile.reset();  // to prevent data race on vptr
```

Comment:
I would prefer to have a comment that explains _why_ otherwise we could have a race.
E.g. "Unregister the query from the query list, otherwise it is still accessible via this list while the query is being destructed, which can result in a data race on the vptr"

Reply:
Done